src/index/index.c

```diff
@@ -1,173 +1,316 @@
- * that possess the provided label and property. */
-Index* Index_Create(Graph *g, const char *label, int label_id, const char *attr_str, Attribute_ID attr_id) {
-  const GrB_Matrix label_matrix = Graph_GetLabelMatrix(g, label_id);
-  GxB_MatrixTupleIter *it;
-  GxB_MatrixTupleIter_new(&it, label_matrix);
-
-  Index *index = rm_malloc(sizeof(Index));
-
-  index->label = rm_strdup(label);
-  index->attribute = rm_strdup(attr_str);
-  index->attr_id = attr_id;
-
-  initializeSkiplists(index);
-
-  Node node;
-  EntityProperty *prop;
-
-  skiplist *sl;
-  NodeID node_id;
-  GraphEntity *entity;
-
-  int found;
-  int prop_index = 0;
-
-  while(true) {
-    bool depleted = false;
-    GxB_MatrixTupleIter_next(it, NULL, &node_id, &depleted);
-    if(depleted) break;
-    Graph_GetNode(g, node_id, &node);
-    // If the sought property is at a different offset than it occupied in the previous node,
-    // then seek and update
-    if (attr_id != ENTITY_PROPS(&node)[prop_index].id) {
-      found = 0;
-      for (int i = 0; i < ENTITY_PROP_COUNT(&node); i ++) {
-        prop = ENTITY_PROPS(&node) + i;
-        if (attr_id == prop->id) {
-          prop_index = i;
-          found = 1;
-          break;
+// Removes fields from index.
+void Index_RemoveField
+(
+    Index *idx,
+    const char *field
+)
+{
+    assert(idx);
+    if(!Index_ContainsField(idx, field)) return;
+
+    for(uint i = 0; i < idx->fields_count; i++) {
+        if(idx->fields[i] == field) {
+            idx->fields_count--;
+            rm_free(idx->fields[i]);
+            array_del_fast(idx->fields, i);
+            array_del_fast(idx->fields_ids, i);
+            break;
         }
-      }
-    } else {
-      found = 1;
     }
-    // The targeted property does not exist on this node
-    if (!found) continue;
+}
 
-    prop = ENTITY_PROPS(&node) + prop_index;
-    // This value will be cloned within the skiplistInsert routine if necessary
-    SIValue *key = &prop->value;
+void Index_IndexNode
+(
+    Index *idx,
+    Node *n
+)
+{
+    double score = 0;           // Default score.
+    const char* lang = NULL;    // Default language.
+    RSIndex *rsIdx = idx->idx;
+    NodeID node_id = ENTITY_GET_ID(n);
 
-    sl = _select_skiplist(index, key->type);
-    if (!sl) continue; // Value was of a type not supported by indices.
-    skiplistInsert(sl, key, node_id);
-  }
+    // Create a document out of node.
+    RSDoc *doc = RediSearch_CreateDocument(&node_id, sizeof(EntityID), score, lang);
 
-  GxB_MatrixTupleIter_free(it);
+    // Add document field for each indexed property.
+    for(uint i = 0; i < idx->fields_count; i++) {
+        if(idx->fields_ids[i] == ATTRIBUTE_NOTFOUND) continue;
 
-  return index;
-}
+        SIValue *v = GraphEntity_GetProperty((GraphEntity*)n, idx->fields_ids[i]);
+        if(v == PROPERTY_NOTFOUND) continue;
+        if(idx->type == IDX_FULLTEXT) {
+            // Value must be of type string.
+            if(v->type != T_STRING) continue;
```

Comment:
I'm assuming this means that if we make a full-text index on a property, any entity with a non-string value for that property will be skipped by index scans. I think this is not ideal, but for the moment it's probably adequate to just document it.

Reply:
Do you think we should get the string representation of numeric values and indexed them as such?